### PR TITLE
feat: add proxy for invoice generation

### LINKED ISSUE #9: Use proxy in requestInvoice
> here we probably also have to use the proxy. 
> 
> similar to `fetch()` - if the proxy should be used do the request through the proxy. 
> 
> so `if (this.options.proxy) {` 
> 
> and for that we need a new endpoint in the proxy app: https://github.com/getAlby/lightning-address-details-proxy/blob/master/main.go
> 
> that endpoint accepts a lightning address and the LNURL params. It then fetches the callback URL and does a request to that callback URL with all other params and simply returns the response.

_Originally posted by @bumi in https://github.com/getAlby/alby-tools/pull/4#discussion_r1122814940_

### PR DESCRIPTION
**Pull request by im-adithya**
Uses https://github.com/getAlby/lightning-address-details-proxy/pull/1

Fixes #9

### CODE REVIEW

File: src/lightning-address.ts [hunk 1]
```diff
@@ -71,14 +71,22 @@ export default class LightningAddress {
...
+    let data;
+    if (this.options.proxy) {
+      const invoiceResult = await fetch(`${this.options.proxy}/generate-invoice?${new URLSearchParams({ ln: this.address, ...params }).toString()}`);
+      const json = await invoiceResult.json();
+      data = json.invoice;
+    } else {
+      const invoiceResult = await fetch(url);
+      data = await invoiceResult.json();
+    }
+
```
Comment (rolznz) [inline src/lightning-address.ts:84]: Do we want to check the response code and add any error handling here?

**Review by rolznz — APPROVED** [review]

File: src/lightning-address.ts [hunk 2]
```diff
@@ -71,14 +71,22 @@ export default class LightningAddress {
     return `https://${this.domain}/.well-known/keysend/${this.username}`;
   }
 
-  async generateInvoice(url: URL): Promise<Invoice> {
-    const data = await fetch(url);
-    const json = await data.json();
-    const paymentRequest = json && json.pr && json.pr.toString();
+  async generateInvoice(url: URL, params: { amount: string, comment?: string, nostr?: string }): Promise<Invoice> {
```
Comment (bumi) [inline src/lightning-address.ts:74]: we should have this generic and I also need to be able to create invoices that have payerdata etc. 

let the user just pass in anything

File: src/lightning-address.ts [hunk 3]
```diff
@@ -101,7 +109,7 @@ export default class LightningAddress {
     let callbackUrl = new URL(callback)
     callbackUrl.search = new URLSearchParams(invoiceParams).toString()
 
-    return this.generateInvoice(callbackUrl);
+    return this.generateInvoice(callbackUrl, invoiceParams);
```
Comment (bumi) [inline src/lightning-address.ts:112]: actually I think we should not pass on the URL here
the object/this knows about the callback URL so we only need to pass the invoiceParams and `generateInvoice` can use the callback URL from `this`

File: src/lightning-address.ts [hunk 4]
```diff
@@ -71,14 +71,22 @@ export default class LightningAddress {
     return `https://${this.domain}/.well-known/keysend/${this.username}`;
   }
 
-  async generateInvoice(url: URL): Promise<Invoice> {
```
Comment (bumi) [inline src/lightning-address.ts:74]: the callback URL here should be used from `this`

**Review by bumi — CHANGES_REQUESTED** [review]

File: src/lightning-address.ts (hunk 4 above)
Comment (im-adithya) [reply src/lightning-address.ts:74]: Done!

**Pull request merged**